File: origin/cell_element_sum.m
function[Gsum,Csum]=cell_element_sum(node,Gc,Cc,dx)
%node = matrix(1,5) elements need calculating sum in matrix x
%x = cell(1,N) in which each cell contain a vector
%dx = matrix(1,N)

Gsum = 0;   Csum = 0;
for i = 1:5,
    if node(i)>0,
        vt1 = Gc{node(i)};
        vt2 = Cc{node(i)};
        Nxi = length(vt1);
        if (mod(i,2)==0)|(i==5),
            n = 1;
        else
            n = Nxi;
        end
        Gsum = Gsum+0.5*vt1(n)*dx(node(i));
                
        Csum = Csum+0.5*vt2(n)*dx(node(i));
    end
end

File: cell_element_sum.py
import numpy as np

def cell_element_sum(node, Gc, Cc, dx):

    node = np.asarray(node, dtype=int)
    dx = np.asarray(dx, dtype=float)
    Gsum = 0.0
    Csum = 0.0
    
    for i in range(1, 6):              # i từ 1 đến 5 như MATLAB
        idx = i - 1                    # index Python: 0-4
        if node[idx] > 0:
            el = node[idx]             # chỉ số phần tử (1-based)
            el_idx = el - 1            # chuyển sang index Python của list
            
            vt1 = np.asarray(Gc[el_idx]).flatten()  # vector G của phần tử này
            vt2 = np.asarray(Cc[el_idx]).flatten()  # vector C
            
            Nxi = len(vt1)
            
            if (i % 2 == 0) or (i == 5):
                n = 0                  # lấy đầu tiên (gần nút bên kia)
            else:
                n = Nxi - 1            # lấy cuối cùng (gần nút hiện tại)
                
            factor = 0.5 * dx[el_idx]
            Gsum += factor * vt1[n]
            Csum += factor * vt2[n]
            
    return Gsum, Csum